Don't add null package to path.

croesch opened the pull request:
## Summary

The path now doesn't get null appended if you provide a scope without a `packageName`.

## Details

If you don't specify the packageName, null gets appended which later leads to a check against a non existing package and then the code defaults to use all classes. With that behavior you effectively can't use the scope without giving a package.

## Context

Suppose you'd like to check all classes, I'd not add a package-name. But I like the ability to specify the scope of the classes to check. With the current behavior this is not possible because null gets added to the path which makes the path invalid.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I've added tests for my code.

What about tests? I'd like to add a test, but the current tests mix test-classes and test-classes of the AUT. I'd consider restructuring src/test/ before adding a test for this problem, so that you can separate test classes from test classes of the AUT.

If not, would you like to make the broken method package private to be able to test it directly?

I was unsure which of both you prefer (or if you have a third possibility) -just give me a hint and I can add a test. Thx

- [ ] Documentation has been updated accordingly to this PR

## Related issue : 

No issue exists for this as far as I can see.

vincent-fuchs commented:
Thanks for the contribution !

I agree it seems there's a problem, but I am not able to reproduce a failure with a test. I've added one with a null package :

```java
   @Test
    public void shouldExecuteConfigurableRuleWithNoPackageProvided_OnAllClasses() {

        ApplyOn applyOn = new ApplyOn(null,"test");

        configurableRule.setRule(DummyCustomRule.class.getName());
        configurableRule.setApplyOn(applyOn);
        configurableRule.setChecks(Arrays.asList("annotatedWithTest"));

        String errorMessage = ruleInvokerService.invokeConfigurableRules(configurableRule, "./target/test-classes/com/societegenerale/commons/plugin/rules/classesForTests");
        assertThat(errorMessage).isNotEmpty();
        assertThat(errorMessage).contains("Architecture Violation");
        assertThat(errorMessage).contains("classes should be annotated with @Test");
    }
```

I am not sure why, but it doesn't prevent all classes to be found, even though we have :

    packageOnRuleToApply=/test-classes/null

during execution

Can you take that test and see if I missed something (I did this a bit quickly so it's possible I missed something) ? I would prefer that we have proof something is broken before fixing it ;-)

croesch commented:
Thanks, the problem with the `null` value arises in `ArchUtils`, there `if (classesPath.toFile().exists()) {` is then `false` and thus the second argument `classFolder` gets ignored.

I also was not able to reproduce it easily with the test suite. I have several ideas about restructuring the test suite. IMO the test suite should be restructured because the above mentioned `if-condition` always is false because `classes` or `test-classes` gets appended to the package which never exists. So the test suite always uses the path that is actually the source of my problem.

I added two files and maven plugin executions to copy class files to a distinct location and call the code under test like in a real world scenario (pointing to `target`). The added test now fails if the code is not changed like suggested.

I would assume that you don't want to merge the change like this, right? So my question would be, do you agree to restructure the tests? If yes, which option would you prefer:

- finish the suggested option with the class-file copying in a clean way
- create different Maven modules (one for the plugin and one for the AUT containing the tests)
- skip adding the test but agree merging the fix ;-)
- any other solution!?

croesch commented:
Sorry for force pushing - I just restored the original commit and added the both commits from yesterday.

vincent-fuchs commented:
Sorry, I am on leave this week, I'll be able to look into it next week-end. Hope it's OK !

croesch commented:
I just reworked the test situation a bit. IMO it now supports running the arch-unit-maven-plugin just against the piece of code we want it to run on.

vincent-fuchs commented:
can you rename _AUT_ to _ApplicationUnderTest_ where it makes sense please ? I was not familiar with the naming and had to google it, so it may not be clear for everyone (I was familiar with "SUT", ie _System Under Test_ so I was pretty sure it was something similar)

vincent-fuchs commented:
I like the way you did it with Maven resources plugin on process-test-classes phase. Let's go this way. You'll need to rebase the PR though, as I merged your other PR and now there's a conflict.

croesch commented:
Okay sorry for the abbreviation. I added a package javadoc explaining it and removed the unnecessary `AUT` and `AUTTest` class. I would prefer the shorter package name but can surely rename it, too.

I rebased and squashed the commits so it should be easier to review the code now.

vincent-fuchs commented:
Thanks for the rebase / squash - it indeed made it easier to review.

While reviewing I found something strange in RuleInvokerServiceTest / RuleInvokerService - I'll create a separate issue
